doc/vidalia.1.in
.TH VIDALIA "1" "June 2006" "@VERSION@"

.SH NAME
vidalia \- Controller GUI for tor 

.SH SYNOPSIS
.B vidalia
.RI [ options ]
.br

.SH DESCRIPTION
Vidalia allows you to start and stop Tor, view the status of Tor at a
glance, and monitor Tor's bandwidth usage. Vidalia also makes it easy to
contribute to the Tor network by helping you setup a Tor server, if you wish.
To find more information about what Tor is and how to use it, see the Tor
homepage: https://www.torproject.org/
.SH OPTIONS
\fB\-help\fR
Displays this usage message and exits.
.TP
\fB\-reset\fR
Resets ALL stored Vidalia settings.
.TP
\fB\-datadir \fI<directory>\fR
Sets Vidalia's datadir to \fI<directory>\fR
.TP
\fB\-pidfile \fI<directory/pidfile>\fR
Sets Vidalia's pidfile to \fI<directory/pidfile>\fR
.TP
\fB\-logfile \fI<file>\fR
Sets the name and location of Vidalia's log file. If this option is specified
without a \fIloglevel\fR, the default level of 'notice' will be used.
.TP
\fB\-loglevel \fI[debug|info|notice|warn|error]\fR
Sets the verbosity of Vidalia's logging. If this option is specified without a 
\fIlogfile\fR, the log messages will be printed to stdout.
.TP
\fB\-style \fI<directory>\fR
Sets Vidalia's interface style. [Windows|Motif|CDE|Plastique]
.TP
\fB\-lang \fI<langcode>\fR
Sets Vidalia's language. Run Vidalia with the \fI-help\fR option to see a list
of available language codes.

.SH SEE ALSO
.BR tor (1)

.BR http://www.torproject.org/projects/vidalia.html

.SH AUTHORS
Matt Edman <edmanm@vidalia-project.net>
.br
Justin Hipple <hipplej@vidalia-project.net>

This manpage was written by Alexander Wirt <formorer@debian.org> for the
Debian Distribution (but may be used by others).



CHANGELOG
0.3.0   06-May-2011
  o Vidalia has got a new GUI. Instead of separate dialogs, each functionality
    is organized in tabs arranged in a common main window. This new tab
    organization will give Vidalia a generic way of organizing the GUI plug-ins
    that will be available in later releases. Resolves bug 2939.
  o When a Tor instance is already running and Vidalia doesn't know the control
    password, don't ask for the it but rather explain the situation and display
    the few possible choices the user has. Resolves bug 2132.
  o Add an option for setting up a non-exit relay to the Sharing configuration
    panel. This is meant to clarify what an exit policy and an exit relay are.
    Resolves bug 2644.
  o Add a way to reload Tor's configuration without having to stop it. Tor can
    reload its configuration while it is running, Vidalia now provides a menu
    option for that, so, for example, relay operators won't be affected by the
    fact that their relay was down for a while. Resolves bug 2724.
  o Reintegrate Breakpad, and make available in other platforms other than
    Windows. Resolves bug 2105.
  o Fix bandwidth assigned to relays on the Network Map. A lot of relays are
    displaying an erroneous bandwidth and since they are ordered by that value
    in the Network Map, it leads to confusion. Vidalia now specifies the
    bandwidth as the minimum of the three possible values (burst, average and
    observed). Fixes bug 2744.
  o Minor change to the checkbox for starting Tor when Vidalia starts. It was
    suggested that the way the phrasing was done was missleading. Resolves bug
    2806.
  o Add a way to bootstrap Tor's torrc file (copy the torrc to a given directory
    before Vidalia starts) so that packages such as Bridge-by-default portable
    bundles for OSX don't violate the directory structure of the operative
    system. Fixes bug 2821.
  o Add the proper CA Certificates so that the "Find Bridges" button works
    again. Fixes bug 2835.
  o Update the useful links help page. Fixes bug 2809.

0.2.12  10-Apr-2011
  o Vidalia's SVN repository has been migrated to Git. All branches but
    master have been archived for later review, since SVN trunk had changed
    significantly; they should be reviewed later to determine whether
    they can and should still be merged. All \version $Id$ headers have been 
    removed since Git does not support $Id$.
  o As part of the move, Vidalia's Trac is now at:
      https://trac.torproject.org/
    All Trac numbers in Vidalia 0.2.12 and beyond refer to the new Trac
    entries. The old Trac is archived for posterity at:
      https://trac-vidalia.torproject.org/projects/vidalia
  o Add support for Tor's ControlSocket as an alternative to ControlPort. It 
    can be used for Linux maintainers to build a better default interaction 
    between Tor and Vidalia by just setting the right permissions and file 
    owner on the socket file for the connection. Using ControlSocket means 
    you don't need to worry about authentication methods with ControlPort. 
    Resolves bug 2091.
  o Add a way to edit arbitrary torrc entries while Tor is running. Now
    Vidalia users have more flexibility for configuring Tor. This change
    doesn't replace editing torrc directly, because on some systems
    (like Debian) Tor can't write to its torrc file. Resolves bug 2083.
  o Remove Vidalia's direct dependency on OpenSSL. This dependency had
    caused Vidalia to fail to run on FreeBSD (due to a bug in the FreeBSD
    ports collection) and Fedora 14 (due to an incompatibility between
    OpenSSL and Fedora's SELinux configuration). Resolves bug 2287 and
    2611.
  o Restore compatibility with Windows 2000. An update to the MiniUPnPc
    library had introduced an unnecessary dependency on a system library
    not included in Windows 2000. Fixes bug 2612.
  o Fix how the advanced message log window displays message updates when
    messages are coming in too quickly, for example when you're listening
    to debug-level messages from Tor. Fixes bug 2093.
  o Add a what's this? link to the bridge option to explain in a more verbose 
    fashion what being a bridge involves. Resolves bug 1995.
  o Prompt users to restart Tor after changing the path to torrc. Fixes bug
    2086.
  o Disable the directory port configuration field when configuring a
    bridge. A bridge does not need to operate a separate directory port,
    and operating one can make a bridge easier to detect. Fixes bug 2431.
  o When Vidalia asks Tor for a bridge's usage history before anyone has
    used it, correctly report that no clients have used the bridge recently.
    Previously, it would incorrectly warn that it was unable to retrieve the
    bridge's usage history. Fixes bug 2186.

0.2.10  26-Aug-2010
  o Drop remote GeoIP lookups. Instead, the default behavior now is to use
    the country-level GeoIP database that ships with Tor to map an IP
    address to a country code, and then map the country code to latitude
    and longitude with a separate database built into Vidalia.
  o Add a -DUSE_GEOIP build option to enable building with MaxMind's GeoIP
    C library for using a local city-level or country-level database
    instead of Tor's database. See README.geoip for details on use.
  o Only update a stream's displayed target address in the network map if
    no hostname was given in the stream's NEW status event. Fix suggested
    by Robert Hogan. (Ticket #608)
  o Update the menubar icon at the same time as the dock icon on OS X.
    Previously, we had a blank icon in the menubar. (Ticket #610)
  o Updated several translations.


0.2.9   20-May-2010
  o Remove the GoDaddy CA certificate bundle since we changed the
    certificate used to authenticate connections to
    geoips.vidalia-project.net for downloading GeoIP information from a
    commercial GoDaddy certificate to a free CACert certificate.

src/vidalia/Vidalia.cpp
/*
**  This file is part of Vidalia, and is subject to the license terms in the
**  LICENSE file, found in the top level directory of this distribution. If you
**  did not receive the LICENSE file with this file, you may obtain it from the
**  Vidalia source package distributed by the Vidalia Project at
**  http://www.torproject.org/projects/vidalia.html. No part of Vidalia, 
**  including this file, may be copied, modified, propagated, or distributed 
**  except according to the terms described in the LICENSE file.
*/

/*
** \file Vidalia.cpp
** \brief Main Vidalia QApplication object
*/

#include "config.h"
#include "Vidalia.h"
#include "LanguageSupport.h"
#include "VMessageBox.h"

#include "stringutil.h"
#include "html.h"

#ifdef USE_MARBLE
#include <MarbleDirs.h>
#endif

#include <QDir>
#include <QTimer>
#include <QTextStream>
#include <QStyleFactory>
#include <QShortcut>
#include <QTranslator>
#include <QLibraryInfo>
#include <QSslSocket>

#ifdef Q_OS_MACX
#include <Carbon/Carbon.h>
#endif
#include <stdlib.h>

/* Available command-line arguments. */
#define ARG_LANGUAGE   "lang"     /**< Argument specifying language.    */
#define ARG_GUISTYLE   "style"    /**< Argument specfying GUI style.    */
#define ARG_RESET      "reset"    /**< Reset Vidalia's saved settings.  */
#define ARG_HELP       "help"     /**< Display usage informatino.       */
#define ARG_DATADIR    "datadir"  /**< Directory to use for data files. */
#define ARG_PIDFILE    "pidfile"  /**< Location and name of our pidfile.*/
#define ARG_LOGFILE    "logfile"  /**< Location of our logfile.         */
#define ARG_LOGLEVEL   "loglevel" /**< Log verbosity.                   */
#define ARG_READ_PASSWORD_FROM_STDIN  \
  "read-password-from-stdin" /**< Read password from stdin. */

/* Static member variables */
QMap<QString, QString> Vidalia::_args; /**< List of command-line arguments.  */
QString Vidalia::_style;               /**< The current GUI style.           */
QString Vidalia::_language;            /**< The current language.            */
TorControl* Vidalia::_torControl = 0;  /**< Main TorControl object.          */
Log Vidalia::_log;
QList<QTranslator *> Vidalia::_translators;


/** Catches debugging messages from Qt and sends them to Vidalia's logs. If Qt
 * emits a QtFatalMsg, we will write the message to the log and then abort().
 */
void
Vidalia::qt_msg_handler(QtMsgType type, const char *s)
{
  QString msg(s);
  switch (type) {
    case QtDebugMsg:
      vDebug("QtDebugMsg: %1").arg(msg);
      break;
    case QtWarningMsg:
      vNotice("QtWarningMsg: %1").arg(msg);
      break;
    case QtCriticalMsg:
      vWarn("QtCriticalMsg: %1").arg(msg);
      break;
    case QtFatalMsg:
      vError("QtFatalMsg: %1").arg(msg);
      break;
  }
  if (type == QtFatalMsg) {
    vError("Fatal Qt error. Aborting.");
    abort();
  }
}

/** Constructor. Parses the command-line arguments, resets Vidalia's
 * configuration (if requested), and sets up the GUI style and language
 * translation. */
Vidalia::Vidalia(QStringList args, int &argc, char **argv)
: QApplication(argc, argv)
{
  qInstallMsgHandler(qt_msg_handler);

  /* Read in all our command-line arguments. */
  parseArguments(args);

  /* Check if we're supposed to reset our config before proceeding. */
  if (_args.contains(ARG_RESET))
    VidaliaSettings::reset();

  /* See if we should load a default configuration file. */
  if (! VidaliaSettings::settingsFileExists())
    copyDefaultSettingsFile();

  /* Handle the -loglevel and -logfile options. */
  if (_args.contains(ARG_LOGFILE))
    _log.open(_args.value(ARG_LOGFILE));
  if (_args.contains(ARG_LOGLEVEL)) {
    _log.setLogLevel(Log::stringToLogLevel(
                      _args.value(ARG_LOGLEVEL)));
    if (!_args.contains(ARG_LOGFILE))
      _log.open(stdout);
  }
  if (!_args.contains(ARG_LOGLEVEL) && 
      !_args.contains(ARG_LOGFILE))
    _log.setLogLevel(Log::Off);

  /* Translate the GUI to the appropriate language. */
  setLanguage(_args.value(ARG_LANGUAGE));
  /* Set the GUI style appropriately. */
  setStyle(_args.value(ARG_GUISTYLE));

  /* Creates a TorControl object, used to talk to Tor. */
  _torControl = new TorControl(TorSettings().getControlMethod());

  /* If we were built with QSslSocket support, then populate the default
   * CA certificate store. */
  loadDefaultCaCertificates();

#ifdef USE_MARBLE
  /* Tell Marble where to stash its generated data */
  Marble::MarbleDirs::setMarbleDataPath(dataDirectory());

#ifdef Q_OS_WIN32
  Marble::MarbleDirs::setMarblePluginPath(vApp->applicationDirPath() 
                                            + "/plugins/marble");
#endif
#endif
#ifdef Q_WS_MAC
  setStyleSheet("QTreeWidget { font-size: 12pt }");
#endif
}

/** Destructor */
Vidalia::~Vidalia()
{
  delete _torControl;
}

/** Enters the main event loop and waits until exit() is called. The signal
 * running() will be emitted when the event loop has started. */
int
Vidalia::run()
{
  QTimer::singleShot(0, vApp, SLOT(onEventLoopStarted()));
  return vApp->exec();
}

/** Called when the application's main event loop has started. This method
 * will emit the running() signal to indicate that the application's event
 * loop is running. */
void
Vidalia::onEventLoopStarted()
{
  emit running();
}

#if defined(Q_OS_WIN)
/** On Windows, we need to catch the WM_QUERYENDSESSION message
 * so we know that it is time to shutdown. */
bool
Vidalia::winEventFilter(MSG *msg, long *result)
{
  if (msg->message == WM_QUERYENDSESSION) {
    quit();
  }
  return QApplication::winEventFilter(msg, result);
}
#endif

/** Returns true if the user wants to see usage information. */
bool
Vidalia::showUsage()
{
  return _args.contains(ARG_HELP);
}

/** Displays usage information for command-line args. */
void
Vidalia::showUsageMessageBox()
{
  QString usage;
  QTextStream out(&usage);

  out << "Available Options:" << endl;
  out << "<table>";
  out << trow(tcol("-"ARG_HELP) + 
              tcol(tr("Displays this usage message and exits.")));
  out << trow(tcol("-"ARG_RESET) +
              tcol(tr("Resets ALL stored Vidalia settings.")));
  out << trow(tcol("-"ARG_DATADIR" &lt;dir&gt;") +
              tcol(tr("Sets the directory Vidalia uses for data files.")));
  out << trow(tcol("-"ARG_PIDFILE" &lt;file&gt;") +
              tcol(tr("Sets the name and location of Vidalia's pidfile.")));
  out << trow(tcol("-"ARG_LOGFILE" &lt;file&gt;") +
              tcol(tr("Sets the name and location of Vidalia's logfile.")));
  out << trow(tcol("-"ARG_LOGLEVEL" &lt;level&gt;") +
              tcol(tr("Sets the verbosity of Vidalia's logging.") +
                   "<br>[" + Log::logLevels().join("|") +"]"));
  out << trow(tcol("-"ARG_GUISTYLE" &lt;style&gt;") +
              tcol(tr("Sets Vidalia's interface style.") +
                   "<br>[" + QStyleFactory::keys().join("|") + "]"));
  out << trow(tcol("-"ARG_LANGUAGE" &lt;language&gt;") + 
              tcol(tr("Sets Vidalia's language.") +
                   "<br>[" + LanguageSupport::languageCodes().join("|") + "]"));
  out << "</table>";

  VMessageBox::information(0, 
    tr("Vidalia Usage Information"), usage, VMessageBox::Ok);
}

/** Returns true if the specified argument expects a value. */
bool
Vidalia::argNeedsValue(QString argName)
{
  return (argName == ARG_GUISTYLE ||
          argName == ARG_LANGUAGE ||
          argName == ARG_DATADIR  ||
          argName == ARG_PIDFILE  ||
          argName == ARG_LOGFILE  ||
          argName == ARG_LOGLEVEL);
}

/** Parses the list of command-line arguments for their argument names and
 * values. */
void
Vidalia::parseArguments(QStringList args)
{
  QString arg, value;

  /* Loop through all command-line args/values and put them in a map */
  for (int i = 0; i < args.size(); i++) {
    /* Get the argument name and set a blank value */
    arg   = args.at(i).toLower();
    value = "";

    /* Check if it starts with a - or -- */
    if (arg.startsWith("-")) {
      arg = arg.mid((arg.startsWith("--") ? 2 : 1));
    }
    /* Check if it takes a value and there is one on the command-line */
    if (i < args.size()-1 && argNeedsValue(arg)) {
      value = args.at(++i);
    }
    /* Place this arg/value in the map */
    _args.insert(arg, value);
  }
}

/** Verifies that all specified arguments were valid. */
bool
Vidalia::validateArguments(QString &errmsg)
{
  /* Check for a language that Vidalia recognizes. */
  if (_args.contains(ARG_LANGUAGE) &&
      !LanguageSupport::isValidLanguageCode(_args.value(ARG_LANGUAGE))) {
    errmsg = tr("Invalid language code specified: ") + _args.value(ARG_LANGUAGE);
    return false;
  }
  /* Check for a valid GUI style */
  if (_args.contains(ARG_GUISTYLE) &&
      !QStyleFactory::keys().contains(_args.value(ARG_GUISTYLE),
                                      Qt::CaseInsensitive)) {
    errmsg = tr("Invalid GUI style specified: ") + _args.value(ARG_GUISTYLE);
    return false;
  }
  /* Check for a valid log level */
  if (_args.contains(ARG_LOGLEVEL) &&
      !Log::logLevels().contains(_args.value(ARG_LOGLEVEL))) {
    errmsg = tr("Invalid log level specified: ") + _args.value(ARG_LOGLEVEL);
    return false;
  }
  /* Check for a writable log file */
  if (_args.contains(ARG_LOGFILE) && !_log.isOpen()) {
    errmsg = tr("Unable to open log file '%1': %2")
                           .arg(_args.value(ARG_LOGFILE))
                           .arg(_log.errorString());
    return false;
  }
  return true;
}

/** Sets the translation Vidalia will use. If one was specified on the
 * command-line, we will use that. Otherwise, we'll check to see if one was
 * saved previously. If not, we'll default to one appropriate for the system
 * locale. */
bool
Vidalia::setLanguage(QString languageCode)
{
  /* If the language code is empty, use the previously-saved setting */
  if (languageCode.isEmpty()) {
    VidaliaSettings settings;
    languageCode = settings.getLanguageCode();
  }
  /* Translate into the desired langauge */
  if (retranslateUi(languageCode)) {
    _language = languageCode;
    return true;
  }
  return false;
}

/** Sets the GUI style Vidalia will use. If one was specified on the
 * command-line, we will use that. Otherwise, we'll check to see if one was
 * saved previously. If not, we'll default to one appropriate for the
 * operating system. */
bool
Vidalia::setStyle(QString styleKey)
{
  /* If no style was specified, use the previously-saved setting */
  if (styleKey.isEmpty()) {
    VidaliaSettings settings;
    styleKey = settings.getInterfaceStyle();
  }
  /* Apply the specified GUI style */
  if (QApplication::setStyle(styleKey)) {
    _style = styleKey;
    return true;
  }
  return false;
}

/** Returns the directory Vidalia uses for its data files. */
QString
Vidalia::dataDirectory()
{
  if (_args.contains(ARG_DATADIR)) {
    return _args.value(ARG_DATADIR);
  }
  return defaultDataDirectory();
}

/** Returns the default location of Vidalia's data directory. */
QString
Vidalia::defaultDataDirectory()
{
#if defined(Q_OS_WIN32)
  return (win32_app_data_folder() + "\\Vidalia");
#elif defined(Q_OS_MAC)
  return (QDir::homePath() + "/Library/Vidalia");
#else
  return (QDir::homePath() + "/.vidalia");
#endif
}

/** Returns the location of Vidalia's pid file. */
QString
Vidalia::pidFile()
{
  if (_args.contains(ARG_PIDFILE)) {
    return _args.value(ARG_PIDFILE);
  }
  return QDir::convertSeparators(dataDirectory() + "/vidalia.pid");
}

bool
Vidalia::readPasswordFromStdin()
{
  return _args.contains(ARG_READ_PASSWORD_FROM_STDIN);
}

/** Writes <b>msg</b> with severity <b>level</b> to Vidalia's log. */
Log::LogMessage
Vidalia::log(Log::LogLevel level, QString msg)
{
  return _log.log(level, msg);
}

/** Creates and binds a shortcut such that when <b>key</b> is pressed in
 * <b>sender</b>'s context, <b>receiver</b>'s <b>slot</b> will be called. */
void
Vidalia::createShortcut(const QKeySequence &key, QWidget *sender,
                        QObject *receiver, const char *slot)
{
  QShortcut *s = new QShortcut(key, sender);
  connect(s, SIGNAL(activated()), receiver, slot);
}

/** Creates and binds a shortcut such that when <b>key</b> is pressed in
 * <b>sender</b>'s context, <b>receiver</b>'s <b>slot</b> will be called. */
void
Vidalia::createShortcut(const QString &key, QWidget *sender,
                        QObject *receiver, const char *slot)
{
  createShortcut(QKeySequence(key), sender, receiver, slot);
}

void
Vidalia::removeAllTranslators()
{
  vInfo("Removing all currently installed UI translator objects.");
  foreach (QTranslator *translator, _translators) {
    QApplication::removeTranslator(translator);
    delete translator;
  }
  _translators.clear();
}

bool
Vidalia::retranslateUi(const QString &languageCode)
{
  QTranslator *systemQtTranslator = 0;
  QTranslator *vidaliaQtTranslator = 0;
  QTranslator *vidaliaTranslator = 0;

  if (! LanguageSupport::isValidLanguageCode(languageCode)) {
    vWarn("Invalid language code: %1").arg(languageCode);
    return false;
  }
  if (! languageCode.compare("en", Qt::CaseInsensitive)) {
    vNotice("Resetting UI translation to English default.");
    _language = languageCode;
    removeAllTranslators();
    return true;
  }

  systemQtTranslator = new QTranslator(vApp);
  Q_CHECK_PTR(systemQtTranslator);
  QString qtDir = QLibraryInfo::location(QLibraryInfo::TranslationsPath);
  systemQtTranslator->load(qtDir + "/qt_" + languageCode + ".qm");


  vidaliaQtTranslator = new QTranslator(vApp);
  Q_CHECK_PTR(vidaliaQtTranslator);
  vidaliaQtTranslator->load(":/lang/qt_" + languageCode + ".qm");

  vidaliaTranslator = new QTranslator(vApp);
  Q_CHECK_PTR(vidaliaTranslator);
  if (! vidaliaTranslator->load(":/lang/vidalia_" + languageCode + ".qm"))
    goto err;

  removeAllTranslators();
  vNotice("Changing UI translation from '%1' to '%2'").arg(_language)
                                                      .arg(languageCode);
  _language = languageCode;
  QApplication::installTranslator(systemQtTranslator);
  QApplication::installTranslator(vidaliaQtTranslator);
  QApplication::installTranslator(vidaliaTranslator);
  _translators << systemQtTranslator
               << vidaliaQtTranslator
               << vidaliaTranslator;

  return true;

err:
  vWarn("Unable to set UI translation to '%1'").arg(languageCode);
  if (systemQtTranslator)
    delete systemQtTranslator;
  if (vidaliaQtTranslator)
    delete vidaliaQtTranslator;
  if (vidaliaTranslator)
    delete vidaliaTranslator;
  delete vidaliaTranslator;
  return false;
}

/** Copies a default settings file (if one exists) to Vidalia's data
 * directory. */
void
Vidalia::copyDefaultSettingsFile() const
{
#ifdef Q_OS_MACX
  CFURLRef confUrlRef;
  CFStringRef pathRef;
  const char *path;

  confUrlRef = CFBundleCopyResourceURL(CFBundleGetMainBundle(), 
                                       CFSTR("vidalia"), CFSTR("conf"), NULL);
  if (confUrlRef == NULL)
    return;

  pathRef = CFURLCopyFileSystemPath(confUrlRef, kCFURLPOSIXPathStyle);
  path    = CFStringGetCStringPtr(pathRef, CFStringGetSystemEncoding());

  if (path) {
    QString defaultConfFile = QString::fromLocal8Bit(path);
    QFileInfo fi(defaultConfFile);
    if (fi.exists()) {
      QFileInfo out(VidaliaSettings::settingsFile());
      if (! out.dir().exists())
        out.dir().mkpath(".");
      QFile::copy(defaultConfFile, out.absoluteFilePath());
    }
  }
  CFRelease(confUrlRef);
  CFRelease(pathRef);
#endif
}

void
Vidalia::loadDefaultCaCertificates() const
{
  QSslSocket::setDefaultCaCertificates(QList<QSslCertificate>());

  if (! QSslSocket::addDefaultCaCertificates(":/pki/EquifaxSecureCA.crt"))
    vWarn("Failed to add the Equifax Secure CA certificate to the default CA "
          "certificate database.");
  if (! QSslSocket::addDefaultCaCertificates(":/pki/DigiCertCA.crt"))
    vWarn("Failed to add the DigiCert Global CA certificate to the default CA "
          "certificate database.");
  if (! QSslSocket::addDefaultCaCertificates(":/pki/DigiCertAssuredCA.crt"))
    vWarn("Failed to add the DigiCert Assured CA certificate to the default CA "
          "certificate database.");
  if (! QSslSocket::addDefaultCaCertificates(":/pki/DigiCertHighAssuranceCA.crt"))
    vWarn("Failed to add the DigiCert High Assurance CA certificate to the default CA "
          "certificate database.");
}

